OpenRA.Mods.Common/Traits/Buildings/DockHost.cs

```diff
@@ -0,0 +1,174 @@
+
+		public BitSet<DockType> GetDockType => Info.Type;
+		public bool IsEnabledAndInWorld => !preventDock && !IsTraitDisabled && !self.IsDead && self.IsInWorld;
+		public int ReservationCount => reservedDockClients.Count;
+		public bool CanReserve => ReservationCount < Info.MaxQueueLength;
+		readonly List<DockClientManager> reservedDockClients = new List<DockClientManager>();
+
+		public WPos DockPosition => self.CenterPosition + Info.DockOffset;
+		public int DockWait => Info.DockWait;
+		public WAngle DockAngle => Info.DockAngle;
+		public bool IsDragRequired => Info.IsDragRequired;
+		public WVec DragOffset => Info.DragOffset;
+		public int DragLength => Info.DragLength;
+
+		[Sync]
+		bool preventDock = false;
+
+		[Sync]
+		Actor dockedClientActor = null;
+		DockClientManager dockedClient = null;
+		readonly Func<Actor, DockClientManager, Actor, DockHost, Activity> dockSequence;
+
+		public DockHost(Actor self, DockHostInfo info)
+			: base(info)
+		{
+			this.self = self;
+
+			// HACK: Remove once TiberianSunRefinery is disposed of.
+			dockSequence = self.Trait<Refinery>().DockSequence;
+		}
+
+		/// <summary> Can this client dock at this host. </summary>
+		/// <remarks>
+		/// Does not check <see cref="DockType"/>.
+		/// Does not check if <see cref="IDockClient"/> is enabled.
+		/// Does not check if <see cref="DockClientManager"/> is enabled.
+		/// </remarks>
+		public virtual bool DockingPossible(Actor clientActor, IDockClient client, bool ignoreReservations = false)
+		{
+			return !IsTraitDisabled && clientActor.Owner.IsAlliedWith(self.Owner) && (ignoreReservations || CanReserve || reservedDockClients.Contains(client.DockClientManager));
+		}
+
+		public virtual bool Reserve(Actor self, DockClientManager client)
+		{
+			if (CanReserve && !reservedDockClients.Contains(client))
+			{
+				reservedDockClients.Add(client);
+				client.ReserveHost(self, this);
+				return true;
+			}
+
+			return false;
+		}
+
+		public virtual void Unreserve()
+		{
+			while (reservedDockClients.Count > 0)
+				Unreserve(reservedDockClients[0]);
+		}
+
+		public virtual void Unreserve(DockClientManager client)
+		{
+			if (reservedDockClients.Contains(client))
+			{
+				reservedDockClients.Remove(client);
+				client.UnreserveHost();
+			}
+		}
+
+		public virtual void DockStarted(Actor clientActor, DockClientManager client)
+		{
+			dockedClientActor = clientActor;
+			dockedClient = client;
+		}
+
+		public virtual void DockCompleted()
+		{
+			dockedClientActor = null;
+			dockedClient = null;
+		}
+
+		public virtual void QueueDockSequence(Activity dockOrder, Actor self, DockClientManager client, Actor hostActor, DockHost host)
+		{
+			dockOrder.QueueChild(dockSequence(self, client, hostActor, host));
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			// Client was killed during docking.
+			if (dockedClientActor != null && (dockedClientActor.IsDead || !dockedClientActor.IsInWorld))
+			{
+				dockedClient = null;
+				dockedClientActor = null;
```

Comment:
it can't be called by a DockCompleted. If a unit dies the activity ends with it

Reply:
This would work wouldn't it? It would compile to the same bytecode, it is just easier to read and modify.

```C#
		public void DockCompleted() { DisposeClient(); }

		public void DisposeClient()
		{
			dockedClientActor = null;
			dockedClient = null;
		}

		void ITick.Tick(Actor self)
		{
			// Client was killed during docking
			if (dockedClient != null && !dockedClient.IsAliveAndInWorld)
				DisposeClient();
		}
```

Reply:
if a unit dies while docking DockCompleted will never be called, we we catch it with the tick method

Reply:
I haven't removed the tick method in my above suggestion, I simply moved the null assignments into their own method and called them from them.

Reply:
Tbh I don't think it's worth uninlining. It's really short code

Reply:
It's not just the length, it's abstracting away the Dispose functionality, a future developer may forget that this needs to occur both on Dock completion and during the Tick due to unit dying. By having both parts call the same Dispose method only the one dispose method needs to be altered for any future changes.

Reply:
I just made the Tick method overridable and call DockCompleted();

Reply:
Ok looks good now.